drivers/platform/stm32/stm32_irq.c

```diff
@@ -0,0 +1,267 @@
+ * IN NO EVENT SHALL ANALOG DEVICES BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, INTELLECTUAL PROPERTY RIGHTS, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*******************************************************************************/
+#include <stdlib.h>
+#include <errno.h>
+#include "util.h"
+#include "irq.h"
+#include "stm32_irq.h"
+#include "gpio.h"
+#include "stm32_gpio.h"
+
+/**
+ * @brief STM32 platform specific IRQ platform ops structure
+ */
+const struct irq_platform_ops stm32_irq_platform_ops = {
+	.init = &stm32_irq_ctrl_init,
+	.register_callback = &stm32_irq_register_callback,
+	.unregister = &stm32_irq_unregister,
+	.global_enable = &stm32_irq_global_enable,
+	.global_disable = &stm32_irq_global_disable,
+	.enable = &stm32_irq_enable,
+	.disable = &stm32_irq_disable,
+	.remove = &stm32_irq_ctrl_remove
+};
+
+/**
+* @struct stm32_irq_callback_desc
+* @brief Structure holding the callback functions for STM32 irqs
+* @note  The callback functions are mapped with 'irq_id' structure
+*/
+typedef struct stm32_irq_callback_desc {
+	uint16_t GPIO_Pin;
+	struct callback_desc desc;
+} stm32_irq_callback_desc[NB_INTERRUPTS];
+
+stm32_irq_callback_desc stm32_irq_callback;
+
+/******************************************************************************/
+/************************ Functions Definitions *******************************/
+/******************************************************************************/
+
+/**
+ * @brief STM32 GPIO external interrupt callback function
+ * @param GPIO_Pin - GPIO pin
+ * @return none
+ */
+void HAL_GPIO_EXTI_Callback(uint16_t GPIO_Pin)
+{
+	enum irq_id id;
+
+	/* Invoke the callback function attached to corresponding GPIO pin */
+	for (id = EXTERNAL_INT_ID1; id <= EXTERNAL_INT_ID5; id++) {
+		if (stm32_irq_callback[id].GPIO_Pin == GPIO_Pin) {
+			stm32_irq_callback[id].desc.callback(stm32_irq_callback[id].desc.ctx,
+							     id, NULL);
+			return;
+		}
+	}
+}
+
+/**
+ * @brief Initialize the STM32 interrupt
+ * @param desc - IRQ descriptor
+ * @param param - IRQ initialization parameters
+ * @return SUCCESS in case of success, FAILURE otherwise
+ */
+int32_t stm32_irq_ctrl_init(struct irq_ctrl_desc **desc,
+			    const struct irq_init_param *param)
+{
+	struct irq_ctrl_desc *descriptor;
+	struct stm32_irq_desc *extra;
+	int32_t ret;
+
+	if (!desc || !param)
+		return -EINVAL;
+
+	descriptor = (struct irq_ctrl_desc *)malloc(sizeof(*descriptor));
+	if (!descriptor) {
+		ret = -ENOMEM;
+		goto error;
+	}
+
+	extra = (struct stm32_irq_desc*)malloc(sizeof(*extra));
+	if (!extra) {
+		ret = -ENOMEM;
```

Comment:
same here. no need to `free(extra)`

Reply:
I got your point. I will update the code accordingly.